src/ooo_cpu.cc

```diff
@@ -475,13 +525,39 @@ void O3_CPU::decode_and_dispatch()
     if (DECODE_BUFFER.occupancy == 0)
         return;
 
+    std::size_t available_decode_bandwidth = DECODE_WIDTH;
+
     // dispatch DECODE_WIDTH instructions that have decoded into the ROB
-    uint32_t count_dispatches = 0;
-    while (count_dispatches < DECODE_WIDTH && DECODE_BUFFER.occupancy > 0 && ROB.occupancy < ROB.SIZE &&
+    while (available_decode_bandwidth > 0 && DECODE_BUFFER.occupancy > 0 && ROB.occupancy < ROB.SIZE &&
             (!warmup_complete[cpu] || ((DECODE_BUFFER.entry[DECODE_BUFFER.head].decoded) && (DECODE_BUFFER.entry[DECODE_BUFFER.head].event_cycle < current_core_cycle[cpu]))))
     {
+        ooo_model_instr &db_entry = DECODE_BUFFER.entry[DECODE_BUFFER.head];
+
+	// Search DIB to see if we need to add this instruction
+	dib_t::value_type &dib_hit_check_set = DIB[(db_entry.ip >> LOG2_BLOCK_SIZE) % DIB_SET];
+	auto way = std::find_if(dib_hit_check_set.begin(), dib_hit_check_set.end(), [db_entry](dib_entry_t x){ return x.valid && ((x.addr >> LOG2_BLOCK_SIZE) == (db_entry.ip >> LOG2_BLOCK_SIZE));});
+	if (way == dib_hit_check_set.end())
+	  {
+	    // we have a miss in the DIB, so we need to replace something
+	    // find victim in DIB
+	    dib_t::value_type &dib_set = DIB[db_entry.ip % DIB_SET];
+	    auto way = std::find_if_not(dib_set.begin(), dib_set.end(), [](dib_entry_t x){ return x.valid; }); // search for invalid
```

Comment:
This looks like a bug. The set we are searching here is different than the set searched on L538. Is this intentional?

Reply:
You're correct, this is wrong.  Good bug catch.